### ISSUE TITLE
Ktor 3.0 support

### ISSUE THREAD

**Issue by lexa-diky** [InsertKoinIO/koin#1716]
**Is your feature request related to a problem? Please describe.**
JB recently released Ktor 3.0.0-betta1.  Currently Ktor does not support it, JB changed API. In particular access to monitor now done via `Application` not `Environment`, maybe there is more incompatibilities. 

**Describe the solution you'd like**
Separate `ktor-3` module to have both ktor 2.* and 3.* 

**Target Koin project**
ktor

**Comment by Nek-12** [comment]
The workaround for now is to copy everything from the koin Ktor sources and fix the compile time error

**Comment by lexa-diky** [comment]
If koin goes with single module approach, it will break comp with ktor-2, and I guess bunch of people will not update for a while

**Comment by Nek-12** [comment]
The only updates for the past months have been dependency updates... The whole library is 2 files

**Comment by arnaudgiuliani** [comment]
Yes, great topic coming 👍

**Comment by Nek-12** [comment]
Can you please add ktor 3.0 support first before you refactor, add new stuff, and break everything with the update you described in your article? @arnaudgiuliani 

Changing one line of code, that is, so that we don't wait till Q4 Sep 2024 for this to get fixed

**Comment by arnaudgiuliani** [comment]
something I can setup for 3.6.0 - Q1
Seems that we need to separate both ktor & ktor3

**Comment by FlyingYu-Z** [comment]
Is there any solution now?

**Comment by JoonasC** [comment]
@arnaudgiuliani I would appreciate a solution for this, as I am currently unable to use SSE (server sent events) in my application due to it only being supported in Ktor 3.0.0. Let me know if I can speed up the work, maybe by making a PR?

**Comment by fluxxion82** [comment]
I'm am interested in this as well and will be watching for updates.

**Comment by Nek-12** [comment]
@fluxxion82 Just copy all sources of the Ktor module to your project. This hasn't been addressed in almost a year and the author has no intention of doing so.

**Comment by KevinMeneses** [comment]
it would be nice to get support for this

**Comment by hashir1296** [comment]
Still waiting for this. No solutions of as of yet?

**Comment by Nek-12** [comment]
Like I said, just copy and paste all sources of Koin-ktor module into your project and fix the breaking changes. Already using this approach in prod

**Comment by hashir1296** [comment]
> Like I said, just copy and paste all sources of Koin-ktor module into your project and fix the breaking changes. Already using this approach in prod

Pardon but can you please briefly describe what do you mean by this? Earlier we were using koin as:
install(Koin) {
        slf4jLogger()
        modules(appModule)
    }
    
How do i migrate from this?

**Comment by hashir1296** [comment]
> > Like I said, just copy and paste all sources of Koin-ktor module into your project and fix the breaking changes. Already using this approach in prod
> 
> Pardon but can you please briefly describe what do you mean by this? Earlier we were using koin as:
> 
> ```
> install(Koin) {
>         slf4jLogger()
>         modules(appModule)
>     }
> ```
> 
> How do i migrate from this?

Figured it out, basically 
1. Goto this link: https://github.com/InsertKoinIO/koin/tree/main/projects/ktor/koin-ktor
2. Manually copy/paste all files under koin-ktor package (I copied only KoinPlugin)
3. In your install block, use your koin plugin instead of ktor's

**Warning**: This is a workaround, with this:
1. You will lose out on  future Ktor or Koin updates
2. You’ll need to keep an eye on both the Koin and Ktor projects for official updates that may render this workaround obsolete.

**Comment by chrisjenx** [comment]
Yeah this is now broken on ktor 3.x:
```
java.lang.IncompatibleClassChangeError: Found interface io.ktor.server.routing.Routing, but class was expected
	at org.koin.ktor.ext.RoutingExtKt.getKoin(RoutingExt.kt:74)
```

Will need a ktor3 package, really the impl is basically the same, but they are incompatible due to changes to classes in ktor3

**Comment by chrisjenx** [comment]
PR here: https://github.com/InsertKoinIO/koin/pull/1978

There is jar with the ktor3 target if you need it for now

**Comment by chrisjenx** [comment]
@arnaudgiuliani if no objections, can we fast track to 4.0? Happy to maintain the module if you ping me.

**Comment by valeriyo** [comment]
Ktor 3.0.0 has been released yesterday -- https://github.com/ktorio/ktor/releases/tag/3.0.0

It would be lovely if Koin supported it officially ❤️

**Comment by arnaudgiuliani** [comment]
I could open track for 4.1 in beta version, to play with this yes 👍

**Comment by sausti** [comment]
> I could open track for 4.1 in beta version, to play with this yes 👍 4.0 is in maintenance mode (no more features).

That would be a huge help!

**Comment by Ynnck123** [comment]
@arnaudgiuliani can you by any chance give an estimated date when such a Beta version with Ktor 3.0.0 support will be testable for us? I'm really looking forward to it

**Comment by batesjernigan-taxa** [comment]
Bumping this. I tried using kodeine also instead of koin to see if the support was there and I ran into the same error mentioned above.
```
java.lang.IncompatibleClassChangeError: Found interface io.ktor.server.routing.Routing, but class was expected
```

**Comment by jsixface** [comment]
> Bumping this. I tried using kodeine also instead of koin to see if the support was there and I ran into the same error mentioned above.
> 
> ```
> java.lang.IncompatibleClassChangeError: Found interface io.ktor.server.routing.Routing, but class was expected
> ```

If you try to inject outside of routing in `Application` scope, it might work.

**Comment by batesjernigan-taxa** [comment]
> > Bumping this. I tried using kodeine also instead of koin to see if the support was there and I ran into the same error mentioned above.
> > ```
> > java.lang.IncompatibleClassChangeError: Found interface io.ktor.server.routing.Routing, but class was expected
> > ```
> 
> If you try to inject outside of routing in `Application` scope, it might work.

That actually worked for me! My update was to move the controller injection code out of the routing block. Thank you for the suggestion! 

```kotlin
fun Application.configureRouting() {
    val fooController by inject<FooController>()
    routing {
        route("/foo") {
            get {            
                // ...
            }
        }
    }
}

```
this was the incorrect code I had before
```kotlin
fun Application.configureRouting() {
    routing {
        route("/foo") {
            get {
                val fooController by inject<FooController>()
                // ...
            }
        }
    }
}
```

**Comment by jsixface** [comment]
You're welcome. 
Well, your code was not wrong in the first place. It's just that the KTOR 3.0 has some breaking changes and Koin 4.0 hasn't caught up with it.

**Comment by Galarzaa90** [comment]
Damn, I had **just** changed all my "controller" files from `Application` extensions to `Routing` extensions :(

**Comment by Wavesonics** [comment]
ooph, open for almost a year 😟 has there been any progress on this?

Any work arounds?

**Comment by chrisjenx** [comment]
I posted a jar you can use, also it's super easy to just copy the plugin code and point at ktor3, there are really only 2 like changes.
But yeah I have an open PR to merge in workings, dunno if @arnaudgiuliani needs me to do anything?

**Comment by Jofairden** [comment]
Any update on the fix?

**Comment by arnaudgiuliani** [comment]
ktor 3 PR is merged. 4.1.0-Beta1 will be released soon to help you test that 👍

**Comment by TheReprator** [comment]
@arnaudgiuliani  any update on it, as i am getting this issue with separte ticket, https://github.com/InsertKoinIO/koin/issues/2066

**Comment by sphrak** [comment]
@TheReprator https://github.com/InsertKoinIO/koin/releases/tag/4.1.0-Beta1

**Comment by TheReprator** [comment]
@sphrak i tried with 4.1.0-Beta1 as well, but not working

**Comment by chrisjenx** [comment]
> @sphrak i tried with 4.1.0-Beta1 as well, but not working

Are you using the new package? If there's an issue createa new ticket. I wrote the new module and has been in prod for a few months now

**Comment by TheReprator** [comment]
@chrisjenx my dependency is as follows,

koin-ktor = { module = "io.insert-koin:koin-ktor", version.ref = "koin" }
koin-annotations = { module = "io.insert-koin:koin-annotations", version.ref = "koin_annotations" }
koin-ksp-compiler = { module = "io.insert-koin:koin-ksp-compiler", version.ref = "koin_annotations" }
koin-logger = { module = "io.insert-koin:koin-logger-slf4j", version.ref = "koin" }

is it correct or it should be else?

**Comment by chrisjenx** [comment]
2.x and 3.x are binary incompatible, you need the `io.insert-koin:koin-ktor3` coords

**Comment by junkeritechnepal** [comment]
The solution i found was to pass dependency from Application Scope above...... routing {.....} 

// ApplicationExt.kt file
```
private fun Application.setupPublicRoutes() {
    val userService by inject<UserService>()

    routing {
        route("api/v1") {
            setupUserRoutes(userService)
        }
    }
}
```

// UserRoutes.kt file
```
fun Route.setupUserRoutes(userService: UserService) {
    post(AppConstants.Route.USER_LOGIN) {

        try {
            val requestDto = call.receive<UserLoginReqDto>()
            val result = userService.read(dto = requestDto)
            call.respond(result = result)
        } catch (ex: Exception) {
            call.respondFailure(logMsg = ex.localizedMessage)
        }
    }
```

**Comment by realityexpander** [comment]
> The solution i found was to pass dependency from Application Scope above...... routing {.....}
> 
> // ApplicationExt.kt file
> 
> ```
> private fun Application.setupPublicRoutes() {
>     val userService by inject<UserService>()
> 
>     routing {
>         route("api/v1") {
>             setupUserRoutes(userService)
>         }
>     }
> }
> ```
> 
> // UserRoutes.kt file
> 
> ```
> fun Route.setupUserRoutes(userService: UserService) {
>     post(AppConstants.Route.USER_LOGIN) {
> 
>         try {
>             val requestDto = call.receive<UserLoginReqDto>()
>             val result = userService.read(dto = requestDto)
>             call.respond(result = result)
>         } catch (ex: Exception) {
>             call.respondFailure(logMsg = ex.localizedMessage)
>         }
>     }
> ```

The problem with that approach is that is eagerly instantiates the objects, and all objects must be passed in as parameters, completely negating the DI framework. Yes, this works (kinda) as a workaround, but here is a better solution:

Put this in your Application.kt file. It's a simple extension function, and can be used in all your Routes to perform injection.

I have tested this with Ktor 3.0.3 and Koin 4.0.4
```
// Fix for Koin `inject`  for Ktor 3.0.3 using Koin 4.0.4 in Routes (e.g. in Routing.kt)
inline fun <reified T : Any> Route.inject(
    qualifier: Qualifier? = null,
    noinline parameters: ParametersDefinition? = null
) =
	lazy {
		GlobalContext.getKoinApplicationOrNull()?.koin?.get<T>(qualifier, parameters) ?:
         inject<T>(T::class.java).value // uses org.koin.java.KoinJavaComponent.inject
	}
```

Until there is a fix, this is a drop-in replacement for the existing Koin `.inject` function, and does the exact same thing, except it works.

**Comment by junkeritechnepal** [comment]
Thanks for the suggestion, definitely looks efficient via lazy load
